collect_app/src/main/java/org/odk/collect/android/activities/MainMenuActivity.java

```diff
@@ -251,4 +263,10 @@ public void onCorrectAdminPassword(Action action) {
     public void onIncorrectAdminPassword() {
         ToastUtils.showShortToast(R.string.admin_password_incorrect);
     }
+
+    @Override
+    public void onProjectSwitched() {
+        invalidateOptionsMenu();
```

Comment:
This might be a nice opportunity to use `LiveData` instead of listeners. This would have the advantage of having less calls for `ProjectSettingsDialog` to make when a project is selected. `ProjectSettingsDialog` and `MainMenuActivity` could both share a `CurrentProjectViewModel`:

```kotlin
class CurrentProjectViewModel {

    val currentProject: LiveData<Project>...

    fun selectProject(project: Project)...
}
```

Then in `MainMenuActivity` you can do:

```kotlin
currentProjectViewModel.getCurrentProject().observe(this) { project ->
    this.currentProject = project
    invalidateOptionsMenu()
}
```

What do you think?

Reply:
I was thinking about using live data but decided that it doesn't make much sense here.
1. In `onPrepareOptionsMenu()` I would take the current project from `MainMenuViewModel` but then when it's switched from `CurrentProjectViewModel` what makes it a bit more complex.
2.  There is just one event that triggers this action I think livedata is more useful in more dynamic cases
3. It requires more code but doesn't make it much better in my opinion.

If you are strongly in favor of livedata we can add it to the list of improvements and discuss it later.

Reply:
> In onPrepareOptionsMenu() I would take the current project from MainMenuViewModel but then when it's switched from CurrentProjectViewModel what makes it a bit more complex.

I think it would be pretty much the same right? You would just do `this.currentProject` instead of `viewModel.getCurrentProject()` in the `onPrepareOptionsMenu`.

> There is just one event that triggers this action I think livedata is more useful in more dynamic cases.

Agreed that it's most useful in those cases. Using reactive values like `LiveData` is also beneficial when you have multiple things that deal with those values though. For example, if we wanted the title bar to display the project name, and we went with the current structure we'd need to change the title in two places (initial setup and when the value changes) whereas with `LiveData` we can just have one place (`observe`) that declares what happens with the current project. The way menus work in Android does take some of that away from us though (having to call `invalidateOptionsMenu`).

> It requires more code but doesn't make it much better in my opinion.

I'm not sure if I agree. Right now we need an interface for the listener, the code to set the listener up and an implementation of that listener which could all disappear if we used a view model. Yes we'd have the extra `ViewModel` but its code would be coming from `ProjectSettingsDialog` and `MainMenuViewModel`.

I do see how it's easy to argue "well we make that change when we add something else" but I think the `ViewModel`/`LiveData` pattern is pretty standard and given we're writing feature tests for this work we have the luxury of being able to refactor our "green"/working solutions as we go to anticipate changes. Personally, I'd advocate making this kind of change now rather than waiting, but it's hard to argue that that is definitively a better idea, so I'll let you make the call here.

Reply:
Ok I implemented it using LiveData.
There is one problem I ran into: https://github.com/getodk/collect/pull/4529/commits/3478dacfb6edeb729c1ca1db868eac248e192fd7#diff-8d4bf37be6c4278258a2d1623c25136611a9d483689a18b2a7eb7ca91e667095R16

this is caused as I explained because of the fact that ResetStateRule is called too late. I know you wanted to look into tests and add some fixes so I handled it like that for now but it should be fixed soon.